src/packages/auth/src/authentication/resolvers/forgotten-password-link/base-resolver.ts

```diff
@@ -0,0 +1,165 @@
+		abstract getForgottenPasswordLink(
+			token: string
+		): Promise<AuthenticationBaseEntity<ForgottenPasswordLinkData>>;
+		abstract getForgottenPasswordLinks(
+			userId: string,
+			period: Date
+		): Promise<AuthenticationBaseEntity<ForgottenPasswordLinkData>[]>;
+		abstract createForgottenPasswordLink(
+			userId: string,
+			token: string
+		): Promise<AuthenticationBaseEntity<ForgottenPasswordLinkData>>;
+		abstract sendForgottenPasswordLink(
+			url: URL,
+			ForgottenPasswordLink: AuthenticationBaseEntity<ForgottenPasswordLinkData>
+		): Promise<boolean>;
+		assertPasswordStrength = assertPasswordStrength ?? defaultPasswordStrength;
+
+		public async generateForgottenPasswordLink(username: string, ctx: AuthorizationContext) {
+			// check that the user exists
+			const user = await this.getUser(username);
+
+			// if the user does not exist, silently fail
+			if (!user?.id) {
+				logger.warn(`User with username ${username} does not exist or is not active.`);
+				return;
+			}
+
+			// Check if the user created X links in the last X period
+			const { rate } = config;
+
+			// which is greater than 24hrs from now
+			const period = new Date(new Date().getTime() - ms(rate.period));
+			const links = await this.getForgottenPasswordLinks(user.id, period);
+
+			// Check rate limiting conditions for forgotten password link creation
+			if (links.length >= rate.limit) {
+				logger.warn(`Too many forgotten password links created for ${username}.`);
+				return;
+			}
+
+			// Create a forgotten password link and save it to the database
+			const link = await this.createForgottenPasswordLink(user.id, createToken());
+
+			// Get Redirect URL
+			const redirect = new URL(
+				ctx?.redirectUri?.toString() ?? requireEnvironmentVariable('AUTH_BASE_URI')
+			);
+
+			const url = new URL(`${redirect.origin}/auth/reset-password`);
+			url.searchParams.set('redirect_uri', redirect.origin.toString());
+			url.searchParams.set('token', link.data.token);
+
+			return { link, url };
+		}
+
+		@Mutation((returns) => Boolean)
+		async sendResetPasswordLink(
+			@Arg('username', () => String) username: string,
+			@Ctx() ctx: AuthorizationContext
+		): Promise<boolean> {
+			const { url, link } = (await this.generateForgottenPasswordLink(username, ctx)) ?? {};
+
+			// fail silently
+			if (!link || !url) {
+				logger.warn(`Failed to create Forgotten Password Link for user with username ${username}.`);
+				return true;
+			}
+
+			// Send the link
+			await this.sendForgottenPasswordLink(url, link);
+
+			return true;
+		}
+
+		@Mutation((returns) => Boolean)
+		async resetPassword(
+			@Arg('token', () => String) token: string,
+			@Arg('password', () => String) password: string
+		): Promise<boolean> {
+			const link = await this.getForgottenPasswordLink(token);
+
+			if (!link) {
+				logger.warn(`Failed to reset password: E0001: Link not found`);
+				throw new AuthenticationError('Reset Password Failed: E0001.');
```

Comment:
Since we're an open source project, people will be able to tell what these are from outside the solution. I reckon we log the full error but return a generic response.

Reply:
Can you provide an example of what you mean here then I can follow.

Reply:
```suggestion
				throw new AuthenticationError('Reset Password Failed');
```